README.md
# Crypto DNS

This packages translates domain names into crypto wallet addresses, using standard TXT records set by the DNS.

## Motivation

Crypto currency wallet addresses aren't readable or rememberable.
Just like in the "normal" DNS system, where we translate domain names to IP addresses, a system which translates domain names to wallet addresses would be useful.

There are several providers of such translation services like ENS which are using smart contracts for the translation, but using those will result in gas fees, which are currently really high.

Fortunately we already have DNS as a global translation system in place, which we can use. By adding TXT records we are able to associate wallet addresses for one or multiple crypto currencies to our domain or subdomain.
This package is heavily inspired by [this blog post of Mattias Geniar](https://ma.ttias.be/proposal-cryptocurrency-addresses-dns/).

## Installation

Using npm:

```bash
npm install crypto-dns
```

Using yarn:

```bash
yarn add crypto-dns
```

## Example

Lookup single address for a currency with the highest priority:

```typescript
import { lookupOne } from 'crypto-dns';

const lookupSingleAddress = await lookupOne('thirdweb.de', 'ETH');
// -> string
// 0xD982065960f77282eDB555b43B175Cf3A7dAC72d
```

You can also lookup all addresses for a currency (sorted by priority):

```typescript
import { lookupMany } from 'crypto-dns';

const lookupMultipleAddresses = await lookupMany('thirdweb.de', 'ETH');
// -> array
//[
//  {
//    version: 1,
//    priority: 10,
//    currency: 'ETH',
//    address: '0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a',
//  },
//  {
//    version: 1,
//    priority: 10,
//    currency: 'ETH',
//    address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d',
//  },
//  {
//    version: 1,
//    priority: 20,
//    currency: 'ETH',
//    address: '0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a',
//  },
//];
```

To receive addresses for every available currency you can use the `lookup` method:

```typescript
import { lookup } from 'crypto-dns';

const lookupResultForAllCurrencies = await lookup('thirdweb.de');
// -> array
//[
//  {
//    version: 1,
//    priority: 10,
//    currency: 'ETH',
//    address: '0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a',
//  },
//  ...
//  {
//    version: 1,
//    priority: 10,
//    currency: 'MATIC',
//    address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d',
//  },
//];
```

## Features

- Use your domain as alias for your crypto wallet addresses
- Available for every crypto currency
- Works with your main domain and all subdomains
- Use priority to weight wallet addresses

### DNS format

```text
crypto:<formatVersion>:<priority> <currency>:<walletAddress>
```

| Keyword       | Description                                                                                                                             |
| ------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
| formatVersion | Version of the format of the DNS string Currently only `1` is available.                                                                |
| priority      | Specify priority of the entry (3 digits). The lower, the higher priority.                                                               |
| currency      | Symbol of the crypto currency (case-insensitive). Use [CoinMarketCap](https://coinmarketcap.com/de/all/views/all/) to find all symbols. |
| walletAddress | String with your wallet address. The format depends on the currency.                                                                    |

Regex:

```regex
/^crypto:(?<formatVersion>\d):(?<priority>\d{1,3})\s(?<currency>\w+):(?<walletAddress>.*)/
```

Examples:

```text
`crypto:v1:10 eth:0xD982065960f77282eDB555b43B175Cf3A7dAC72d`
`crypto:v1:10 matic:0xD982065960f77282eDB555b43B175Cf3A7dAC72d`
`crypto:v1:10 btc:XXXX`
```

### How to use subdomains

Subdomains can be used like root domains. Just use your subdomain when adding it to the DNS instead of the `@` path.

## Configuration

You can pass a configuration object as the last parameter on each lookup-Function.

```typescript
const config = {
  nameserver: 'https://8.8.8.8/resolve',
  timeout: 1500,
};

const lookupSingleAddress = await lookupOne('thirdweb.de', 'ETH', config);
```

The following configurations are available:

| Configuration | Type   | Default                     | Description                                                                                                |
| ------------- | ------ | --------------------------- | ---------------------------------------------------------------------------------------------------------- |
| `nameserver`  | string | `https://1.1.1.1/dns-query` | Server-URL for DoH-Resolution (Cloudflare: `https://1.1.1.1/dns-query`, Google: `https://8.8.8.8/resolve`) |
| `timeout`     | number | `2000`                      | Timeout for DoH HTTP-Call                                                                                  |

## Tests

You can validate all tests by running:

```bash
yarn install
yarn test
```

## License

This library is licensed under the [**MIT License**](https://github.com/LuckyF/crypto-dns/blob/main/LICENSE.md).

## Bugs & Feature Requests

For Bug reports or feature requests, please [create an issue](https://github.com/LuckyF/crypto-dns/issues) in the crypto-dns repository.

## Support & Funding

If you would like to support or fund the development of this project, feel free to [contact me via mail](mailto:hey@frischknecht.dev?subject=Support%20Crypto-DNS).


CHANGELOG.md
## [1.2.3-rc.4](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.3-rc.3...v1.2.3-rc.4) (2023-01-26)

## [1.2.3-rc.3](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.3-rc.2...v1.2.3-rc.3) (2023-01-02)

## [1.2.3-rc.2](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.3-rc.1...v1.2.3-rc.2) (2022-12-08)

## [1.2.3-rc.1](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.2...v1.2.3-rc.1) (2022-10-18)

## [1.2.2](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.1...v1.2.2) (2022-10-08)

## [1.2.2-rc.1](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.1...v1.2.2-rc.1) (2022-10-08)

## [1.2.1](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.0...v1.2.1) (2022-10-04)

## [1.2.1-rc.1](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.2.0...v1.2.1-rc.1) (2022-10-04)

## [1.2.0](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.1.0...v1.2.0) (2022-08-22)


### ⚠ BREAKING CHANGES

* Migrate to new repository and CICD Pipeline

### Features

* Migrate to new repository and CICD Pipeline ([86d5257](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/commit/86d52574a9cf662731cd8a76bd18fda3134921f0))

## [1.1.0](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/compare/v1.0.0...v1.1.0) (2022-08-22)


### Features

* Migrate to new repository and CICD Pipeline ([5aa0648](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/commit/5aa06480d34ef7553c04bdfca156ff69b4fd1ac3))

## 1.0.0 (2022-08-22)


### Features

* initial commit ([82df851](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/commit/82df8515e2654016a2bca1a92d796f551d323738))


### CI/CD Pipelines

* Fix typo ([dd2659b](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/commit/dd2659bb61f60a2ac2c71edcd1d8ebd55308c3da))
* Fix typo ([289f686](https://git.frischknecht.dev/crypto-dns/packages/crypto-dns/commit/289f686cd837e6d25bb98a436914ea62474e59de))


src/lookup.spec.ts
import axios from 'axios';
import { lookup } from './index';

jest.mock('axios');
describe('Test mocked DNS lookup', () => {
  test('Network outage (axios error)', async () => {
    const data = {
      request: {},
      response: undefined,
      message: 'timeout of XXXms exceeded',
      code: 'ECONNABORTED',
      isAxiosError: true,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.reject(data));
    expect.assertions(1);

    try {
      await lookup('thirdweb.de');
    } catch (error) {
      expect(error.message).toMatch(/Network error :.*/);
    }
  });

  test('DoH request - 400', async () => {
    const data = {
      request: {},
      response: {
        status: 400,
        statusText: 'Bad Request',
      },
      message: 'Request failed with status code 400',
      code: undefined,
      isAxiosError: true,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.reject(data));
    expect.assertions(1);

    try {
      await lookup('thirdweb.de');
    } catch (error) {
      expect(error.message).toMatch(/Network error :.*/);
    }
  });

  test('DoH request - 500', async () => {
    const data = {
      request: {},
      response: {
        status: 500,
        statusText: 'Internal Error',
      },
      message: 'Request failed with status code 500',
      code: undefined,
      isAxiosError: true,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.reject(data));
    expect.assertions(1);

    try {
      await lookup('thirdweb.de');
    } catch (error) {
      expect(error.message).toMatch(/Network error :.*/);
    }
  });

  test('Request error (no DoH server)', async () => {
    const data = {
      data: {},
      status: 200,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));
    expect.assertions(1);

    try {
      await lookup('thirdweb.de');
    } catch (error) {
      expect(error.message).toMatch('Unplausible DoH response');
    }
  });

  test('DoH error', async () => {
    const data = {
      data: {
        Status: 1,
        Answer: [],
      },
      status: 200,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));
    expect.assertions(1);

    try {
      await lookup('thirdweb.de');
    } catch (error) {
      expect(error.message).toMatch('Failed DoH response');
    }
  });

  test('DNS Verification', async () => {
    const data = {
      data: {
        Status: 0,
        TC: false,
        RD: true,
        RA: true,
        AD: false,
        CD: false,
        Question: [{ name: 'thirdweb.de', type: 16 }],
        Answer: [
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:10 btc:bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
          },
          {
            name: 'thirdweb.de',
            type: 46,
            TTL: 300,
            data: 'TXT ECDSAP256SHA256 2 300 1640864421 1640684421 34505 thirdweb.de. IHsOcPmC0fVP3aVofEadAKKpZrUeGuro6UBlRdOBMjdLw1+ekfmdxOymjVpbdmb8hK03WRGUkqTA3ngDMDTs6Q==',
          },
        ],
      },
      status: 200,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));
    expect.assertions(1);

    try {
      await lookup('thirdweb.de');
    } catch (error) {
      expect(error.message).toMatch('Failed DNSSEC validation');
    }
  });

  test('Correct DoH request', async () => {
    const data = {
      data: {
        Status: 0,
        TC: false,
        RD: true,
        RA: true,
        AD: true,
        CD: false,
        Question: [{ name: 'thirdweb.de', type: 16 }],
        Answer: [
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:10 btc:bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
          },
          {
            name: 'thirdweb.de',
            type: 46,
            TTL: 300,
            data: 'TXT ECDSAP256SHA256 2 300 1640864421 1640684421 34505 thirdweb.de. IHsOcPmC0fVP3aVofEadAKKpZrUeGuro6UBlRdOBMjdLw1+ekfmdxOymjVpbdmb8hK03WRGUkqTA3ngDMDTs6Q==',
          },
        ],
      },
      status: 200,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));

    const result = await lookup('thirdweb.de');

    expect(result).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          version: 1,
          priority: 10,
          currency: 'BTC',
          address: 'bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu',
        }),
      ]),
    );
  });
});

describe('Test "lookup" function', () => {
  test('Multiple addresses', async () => {
    const data = {
      data: {
        Status: 0,
        TC: false,
        RD: true,
        RA: true,
        AD: true,
        CD: false,
        Question: [{ name: 'thirdweb.de', type: 16 }],
        Answer: [
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:10 btc:bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
          },
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:10 eth:0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a"',
          },
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:10 eth:0xD982065960f77282eDB555b43B175Cf3A7dAC72d"',
          },
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:20 eth:0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a"',
          },
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:10 matic:0xD982065960f77282eDB555b43B175Cf3A7dAC72d"',
          },
          {
            name: 'thirdweb.de',
            type: 46,
            TTL: 300,
            data: 'TXT ECDSAP256SHA256 2 300 1640864421 1640684421 34505 thirdweb.de. IHsOcPmC0fVP3aVofEadAKKpZrUeGuro6UBlRdOBMjdLw1+ekfmdxOymjVpbdmb8hK03WRGUkqTA3ngDMDTs6Q==',
          },
        ],
      },
      status: 200,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));

    const result = await lookup('thirdweb.de');

    expect(result).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          version: 1,
          priority: 10,
          currency: 'BTC',
          address: 'bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu',
        }),
        expect.objectContaining({
          version: 1,
          priority: 10,
          currency: 'ETH',
          address: '0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a',
        }),
        expect.objectContaining({
          version: 1,
          priority: 10,
          currency: 'ETH',
          address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d',
        }),
        expect.objectContaining({
          version: 1,
          priority: 20,
          currency: 'ETH',
          address: '0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a',
        }),
        expect.objectContaining({
          version: 1,
          priority: 10,
          currency: 'MATIC',
          address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d',
        }),
      ]),
    );
  });

  test('No address', async () => {
    const data = {
      data: {
        Status: 0,
        TC: false,
        RD: true,
        RA: true,
        AD: true,
        CD: false,
        Question: [{ name: 'thirdweb.de', type: 16 }],
        Answer: [],
      },
      status: 200,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));

    const result = await lookup('thirdweb.de');

    expect(result).toEqual([]);
  });

  test('Malformed addresses', async () => {
    const data = {
      data: {
        Status: 0,
        TC: false,
        RD: true,
        RA: true,
        AD: true,
        CD: false,
        Question: [{ name: 'thirdweb.de', type: 16 }],
        Answer: [
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:156 btc;bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
          },
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"crypto:1:11btc;bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
          },
          {
            name: 'thirdweb.de',
            type: 16,
            TTL: 300,
            data: '"coin:1:10btc;bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
          },
          {
            name: 'thirdweb.de',
            type: 46,
            TTL: 300,
            data: 'TXT ECDSAP256SHA256 2 300 1640864421 1640684421 34505 thirdweb.de. IHsOcPmC0fVP3aVofEadAKKpZrUeGuro6UBlRdOBMjdLw1+ekfmdxOymjVpbdmb8hK03WRGUkqTA3ngDMDTs6Q==',
          },
        ],
      },
      status: 200,
    };

    (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));

    const result = await lookup('thirdweb.de');

    expect(result).toEqual([]);
  });
});


src/lookupOne.spec.ts
import axios from 'axios';

import { lookupOne } from './index';

jest.mock('axios');

beforeEach(() => {
  const data = {
    data: {
      Status: 0,
      TC: false,
      RD: true,
      RA: true,
      AD: true,
      CD: false,
      Question: [{ name: 'thirdweb.de', type: 16 }],
      Answer: [
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 btc:bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 eth:0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 eth:0xD982065960f77282eDB555b43B175Cf3A7dAC72d"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:20 eth:0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 matic:0xD982065960f77282eDB555b43B175Cf3A7dAC72d"',
        },
        {
          name: 'thirdweb.de',
          type: 46,
          TTL: 300,
          data: 'TXT ECDSAP256SHA256 2 300 1640864421 1640684421 34505 thirdweb.de. IHsOcPmC0fVP3aVofEadAKKpZrUeGuro6UBlRdOBMjdLw1+ekfmdxOymjVpbdmb8hK03WRGUkqTA3ngDMDTs6Q==',
        },
      ],
    },
    status: 200,
  };

  (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));
});

describe('Test "lookupOne" function', () => {
  test('One of multiple DNS wallets', async () => {
    const result = await lookupOne('thirdweb.de', 'eth');

    expect(result).toMatch(
      /(0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a|0xD982065960f77282eDB555b43B175Cf3A7dAC72d|0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a)/,
    );
  });

  test('Single DNS wallet', async () => {
    const result = await lookupOne('thirdweb.de', 'matic');

    expect(result).toBe('0xD982065960f77282eDB555b43B175Cf3A7dAC72d');
  });

  test('lookupOne - Non existing wallet', async () => {
    const result = await lookupOne('thirdweb', 'sol');

    expect(result).toBe(null);
  });
});


src/lookupMany.spec.ts
import axios from 'axios';
import { lookupMany } from './index';

jest.mock('axios');

beforeEach(() => {
  const data = {
    data: {
      Status: 0,
      TC: false,
      RD: true,
      RA: true,
      AD: true,
      CD: false,
      Question: [{ name: 'thirdweb.de', type: 16 }],
      Answer: [
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 btc:bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 eth:0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 eth:0xD982065960f77282eDB555b43B175Cf3A7dAC72d"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:20 eth:0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a"',
        },
        {
          name: 'thirdweb.de',
          type: 16,
          TTL: 300,
          data: '"crypto:1:10 matic:0xD982065960f77282eDB555b43B175Cf3A7dAC72d"',
        },
        {
          name: 'thirdweb.de',
          type: 46,
          TTL: 300,
          data: 'TXT ECDSAP256SHA256 2 300 1640864421 1640684421 34505 thirdweb.de. IHsOcPmC0fVP3aVofEadAKKpZrUeGuro6UBlRdOBMjdLw1+ekfmdxOymjVpbdmb8hK03WRGUkqTA3ngDMDTs6Q==',
        },
      ],
    },
    status: 200,
  };

  (axios.get as jest.Mock).mockImplementationOnce(() => Promise.resolve(data));
});

describe('Test "lookupMany" function', () => {
  test('Multiple DNS wallets of single currency', async () => {
    const result = await lookupMany('thirdweb.de', 'eth');

    expect(result).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          version: 1,
          priority: 10,
          currency: 'ETH',
          address: '0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a',
        }),
        expect.objectContaining({
          version: 1,
          priority: 10,
          currency: 'ETH',
          address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d',
        }),
        expect.objectContaining({
          version: 1,
          priority: 20,
          currency: 'ETH',
          address: '0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a',
        }),
      ]),
    );
  });

  test('Non existing wallet', async () => {
    const result = await lookupMany('thirdweb.de', 'sol');

    expect(result).toEqual([]);
  });
});


.eslintrc.js
'use strict';

/**
 * @type {import('eslint'.Linter.Config)}
 */
const config = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: { project: './tsconfig.json', tsconfigRootDir: __dirname, sourceType: 'module' },
  plugins: ['@typescript-eslint/eslint-plugin', 'no-loops', 'prettier'],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended',
  ],
  rules: {
    'no-console': 'warn',
    'prettier/prettier': 'error',
    'no-loops/no-loops': 'error',
    quotes: ['error', 'single'],
    '@typescript-eslint/no-floating-promises': 'warn',
  },
  ignorePatterns: ['node_modules', 'dist', 'coverage', '.eslintrc.js', '**/*.config.js', '**/*.config.ts'],
};

module.exports = { ...config };


src/index.ts
import axios, { AxiosResponse } from 'axios';
import { CryptoDNSConfigurationDefaultI, CryptoDNSConfigurationI, CryptoDNSEntryI, DoHEntry } from './types/types';

const defaultConfig: CryptoDNSConfigurationDefaultI = { nameserver: 'https://1.1.1.1/dns-query', timeout: 2000 }; // Google DNS: https://8.8.8.8/resolve

export const lookup = async (
  domain: string,
  config: CryptoDNSConfigurationI = defaultConfig,
): Promise<CryptoDNSEntryI[]> => {
  const mergedConfig = { ...defaultConfig, ...config };

  const result: CryptoDNSEntryI[] = [];
  let dnsResponse: AxiosResponse;

  try {
    dnsResponse = await axios.get(mergedConfig.nameserver, {
      headers: { accept: 'application/dns-json' },
      params: { name: domain, type: 'TXT' },
      //params: { name: domain, type: 'TXT', do: true, cd: false },
      timeout: mergedConfig.timeout,
    });
  } catch (error) {
    throw new Error(`Network error :${error.message}`);
  }

  if (dnsResponse?.data?.Status === undefined || dnsResponse?.data?.Answer === undefined) {
    throw new Error('Unplausible DoH response');
  } else if (dnsResponse.data.Status != 0) {
    throw new Error('Failed DoH response');
  } else if (dnsResponse.data.AD !== true || dnsResponse.data.CD !== false) {
    throw new Error('Failed DNSSEC validation');
  }

  dnsResponse.data.Answer.forEach((entry: DoHEntry) => {
    if (entry.type !== 16) {
      return;
    }

    // Remove " if entry.data is wrapped like "crypto:1:10 matic:0xXYZ"
    if (entry.data.startsWith('"') && entry.data.endsWith('"')) {
      entry.data = entry.data.slice(1, -1);
    }
    const dnsEntry = entry.data.match(
      /^crypto:(?<formatVersion>\d):(?<priority>\d{1,3})\s(?<currency>\w+):(?<walletAddress>.*)/,
    );
    if (!dnsEntry) {
      return;
    }

    result.push({
      version: Number(dnsEntry.groups?.formatVersion),
      priority: Number(dnsEntry.groups?.priority),
      currency: dnsEntry.groups?.currency.toUpperCase() || '',
      address: dnsEntry.groups?.walletAddress || '',
    });
  });

  return result;
};

export const lookupMany = async (
  domain: string,
  currency: string,
  config: CryptoDNSConfigurationI = defaultConfig,
): Promise<CryptoDNSEntryI[]> => {
  // Request addresses from DNS
  const result = await lookup(domain, config);

  return (
    result
      // Filter addresses for one currency
      .filter((element) => {
        return element.currency === currency.toUpperCase();
      })
      // Sort addresses by priority
      .sort((a, b) => {
        return a.priority - b.priority ? 1 : b.priority - a.priority ? -1 : 0;
      })
  );
};

export const lookupOne = async (
  domain: string,
  currency: string,
  config: CryptoDNSConfigurationI = defaultConfig,
): Promise<string | null> => {
  // Request addresses for one currency
  let lookupResult = await lookupMany(domain, currency, config);

  // Return null if no address has been received
  if (!lookupResult.length) {
    return null;
  }

  // Determine highest priority
  const highestPrio = lookupResult[0].priority;
  // Filter for highest priority
  lookupResult = lookupResult.filter((element) => {
    return element.priority === highestPrio;
  });

  // If multiple addresses are set, select one random address
  return lookupResult[Math.floor(Math.random() * lookupResult.length)].address;
};

//const main = async () => {
//  console.log(await lookup('thirdweb.de'));
//  console.log(await lookupMany('thirdweb.de', 'eth'));
//  console.log(await lookupOne('thirdweb.de', 'eth'));
//};
//main();


.releaserc.js
'use strict';

const config = {
  preset: 'angular',
  branches: [
    {
      name: 'main',
    },
    {
      name: 'next',
      channel: 'next',
      prerelease: 'rc',
    },
  ],
  plugins: [
    [
      '@semantic-release/commit-analyzer',
      {
        preset: 'conventionalcommits',
        releaseRules: [
          { type: 'build', release: false },
          { type: 'ci', release: false },
          { type: 'chore', release: false },
          { type: 'chore', scope: 'dependencies', release: 'patch' },
          { type: 'docs', release: false },
          { type: 'docs', scope: 'README', release: 'patch' },
          { type: 'feat', release: 'minor' },
          { type: 'fix', release: 'patch' },
          { type: 'perf', release: 'patch' },
          { type: 'refactor', release: 'patch' },
          { type: 'style', release: false },
          { type: 'test', release: false },
        ],
      },
    ],
    [
      '@semantic-release/release-notes-generator',
      {
        preset: 'conventionalcommits',
        presetConfig: {
          types: [
            { type: 'build', section: 'CI/CD Pipelines', hidden: false },
            { type: 'ci', section: 'CI/CD Pipelines', hidden: false },
            { type: 'chore', section: 'Miscellaneous Chores', hidden: true },
            { type: 'chore', scope: 'dependencies', section: 'Dependency Updates', hidden: false },
            { type: 'docs', section: 'Documentation', hidden: false },
            { type: 'feat', section: 'Features' },
            { type: 'fix', section: 'Bug Fixes' },
            { type: 'perf', section: 'Performance Improvements' },
            { type: 'refactor', section: 'Code Refactoring', hidden: false },
            { type: 'style', section: 'Styles', hidden: false },
            { type: 'test', section: 'Tests', hidden: false },
          ],
        },
      },
    ],
    [
      '@semantic-release/changelog',
      {
        changelogFile: 'CHANGELOG.md',
      },
    ],
    //[
    //  '@semantic-release/exec',
    //  {
    //    verifyReleaseCmd: 'echo ${nextRelease.version} > VERSION',
    //  },
    //],
    [
      '@semantic-release/npm',
      {
        npmPublish: true,
      },
    ],
    [
      'semantic-release-npm-deprecate-old-versions',
      {
        rules: [
          {
            rule: 'supportLatest',
            options: {
              numberOfMajorReleases: 1,
              numberofMinorReleases: 3,
              numberofPatchReleases: 3,
            },
          },
          {
            rule: 'supportPreReleaseIfNotReleased',
            options: {
              numberOfPreReleases: 1,
            },
          },
          {
            rule: 'deprecateAll',
            options: {},
          },
        ],
      },
    ],
    ['@semantic-release/gitlab', {}],
    [
      '@semantic-release/git',
      {
        assets: ['package.json', 'CHANGELOG.md'],
        message: 'chore(release): Release v${nextRelease.version}\n\n${nextRelease.notes}',
      },
    ],
    [
      '@saithodev/semantic-release-backmerge',
      {
        branches: [{ from: 'main', to: 'next' }],
      },
    ],
  ],
};

module.exports = { ...config };


examples/index.js
const { lookup, lookupOne, lookupMany } = require('crypto-dns');

const main = async () => {
  const resultLookup = await lookup('thirdweb.de');
  console.info(`\n\nlookup():`);
  console.log(resultLookup);
  /**
   * Returns:
   * [
   *   {
   *     version: 1,
   *     priority: 10,
   *     currency: 'BTC',
   *     address: 'bc1qt44xtffh368az9s6r4qa3cgf4sqzjq2hk7nneu'
   *   },
   *   {
   *     version: 1,
   *     priority: 10,
   *     currency: 'ETH',
   *     address: '0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a'
   *   },
   *   {
   *     version: 1,
   *     priority: 10,
   *     currency: 'ETH',
   *     address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d'
   *   },
   *   {
   *     version: 1,
   *     priority: 20,
   *     currency: 'ETH',
   *     address: '0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a'
   *   },
   *   {
   *     version: 1,
   *     priority: 10,
   *     currency: 'MATIC',
   *     address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d'
   *   }
   * ]
   */

  const resultLookupOne = await lookupOne('thirdweb.de', 'ETH');
  console.info(`\n\nlookupOne():`);
  console.log(resultLookupOne);
  /**
   * Returns:
   * 0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a
   */

  const resultLookupMany = await lookupMany('thirdweb.de', 'ETH');
  console.info(`\n\nlookupMany():`);
  console.log(resultLookupMany);
  /**
   * Returns:
   * [
   *   {
   *     version: 1,
   *     priority: 10,
   *     currency: 'ETH',
   *     address: '0xB9Af69a9850a98d9Fb66Ce210E88021Ad583961a'
   *   },
   *   {
   *     version: 1,
   *     priority: 10,
   *     currency: 'ETH',
   *     address: '0xD982065960f77282eDB555b43B175Cf3A7dAC72d'
   *   },
   *   {
   *     version: 1,
   *     priority: 20,
   *     currency: 'ETH',
   *     address: '0xccaa72d80EeB1A2Ac91B6Fdebff995D55ea9368a'
   *   }
   * ]
   */
};

main();
